funsor/terms.py

```diff
@@ -65,7 +66,9 @@ def reflect(cls, *args, **kwargs):
         assert len(new_args) == len(cls._ast_fields)
         _, args = args, new_args
 
-    cache_key = tuple(id(arg) if not isinstance(arg, Hashable) else arg for arg in args)
+    # JAX DeviceArray has .__hash__ method but raise the unhashable error there.
```

Comment:
I think the issue is that `__hash__` is defined but raises a `TypeError`. Instead if it set `__hash__` to `None`, Python would raise an appropriate `TypeError`, we should probably fix this in JAX.

Reply:
@neerajprad @eb8680 In JAX, they consider DeviceArray, ndarray as [Hashably](https://github.com/google/jax/blob/master/jax/util.py#L186), which [seems to do](https://github.com/google/jax/blob/master/jax/api.py#L1393) what we do here. I think having `__hash__` returns `TypeError` is their intention.

Reply:
@fehiepsi - This isn't a big deal, it just might be confusing to new users, and I think the pythonic way would be to just set `__hash__ = None`, which will throw a TypeError by default (I don't think the error message is particularly informative that we need to override it). Quoting from the [python docs](https://docs.python.org/3/reference/datamodel.html):

> A class which defines its own __hash__() that explicitly raises a TypeError would be incorrectly identified as hashable by an isinstance(obj, collections.abc.Hashable) call.

Anyways, this is a really minor point, and we can bring it up with the jax devs if needed later.